src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -486,6 +486,17 @@ public String getCookie() {
             return getDisplayName();
         }
 
+        @Override
+        public String getAffinityKey() {
+            String enclosingLabel = getEnclosingLabel();
```

Comment:
That method is designed for UI (like the executor widget) and is not actually what you want here. For example, given

```groovy
node {
  stage('Build') {
    sh 'make'
  }
  stage('Test') {
    sh 'make test'
  }
}
```

the result of `getEnclosingLabel` will vary according to which stage we are currently inside. That is inappropriate for the affinity key, which is supposed to be a stable function of what is _outside_ this `node` block. (After all, the affinity key should be checked just when the `node` block starts.)

So all the machinery with the hash codes and the `match` logic is out of place here, and there is no need to patch `getEnclosingLabel`. You want something much simpler for this purpose: search `executorStepNode.iterateEnclosingBlocks` for some identifier which ought to be stable across runs.

Alternately, just use `ownerTask.name` by default as the affinity key, and if users wish to control the affinity better, let them do so as an explicit parameter to `node`. My comment in https://github.com/jenkinsci/junit-plugin/pull/94#discussion_r155143817 follows similar logic.

Reply:
Thanks for the hint about this. Thinking about it a second time I could have just done a much simpler implementation instead of using `getEnclosingLabel()`.
Initially I was only returning  ownerTask,name. However it turned out that when using parallel blocks it happens to happen quite often that still different nodes are picked up.
Therefore I'll go with something else. In fact I already have an implementation for checking for stages and parallel blocks outside nodes blocks used at some different place. I'll probably use that one.